HA implementation for RouteFlow

srijanmishra opened the pull request:
The corresponding changes consist of 
- RFMonitor implementation that
  - Monitors all the controllers for failure.
  - In case of failure elect a new master and inform rfproxy about the new master.
- updated rftest1 that has two controllers running as master and slave.
- fixes in msgen.py to generate PEP8 conformant code and don't generate unused code.

srijanmishra committed:
RFMonitor init

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Message to register Controller with RFMonitor

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Conform to PEP8

Added 2 newlines between classes. Removed an unused import pymongo.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Register controller with RFMonitor

For communication between controller instance and RFMontor defined a new
channel id and RFMonitor instance id. Added message processor in
RFMonitor that stores the details of controller instances with RFMonitor instance

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Automated ControllerRegister message generation.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer <joestringernz@gmail.com>

srijanmishra committed:
PEP8 conformed code generation by mesgen.py

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Test to check if a controller is alive.

To check if controller is up and running, create a socket and try to
connect it to the listening host and port of the controller. If the
controller is dead, then delete the corresponding entry from RFMonitor
instance.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Handle controller spawn and death

When a controller is spawned its details are registered with RFMonitor
and a `Monitor` is initialized for the controller that is responsible
for testing periodically if controller is alive. When a controller dies,
entries corresponding to the controller are removed from RFMonitor and
periodic test going on the controller is stopped.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Monitor class for scheduling tests on controller.

When the test starts, the monitor continuously tests for controller
failure at time intervals equal to `callback_time` which is currently
equal to 1 second. The class exposes an API to RFMonitor for starting
and stoping tests.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Modified rftest1 for HA

Starts RFMonitor to monitor controller failure and two controller
instances with one as master and other as slave.
Openvswitch connects to both the controllers and sends info only to
master connection.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Docstring for functions.

Docstring for various functions describing the purpose and keyword
arguments passed to the functions.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Set socket used for test to non-blocking.

The socket being used to check if a controller is alive was blocking. It
blocked until it was able to connect to the given host and port.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Code refactor to handle controller failure.

The test function that was responsible for checking if a controller is
alive had the code to amend RFMonitor instance and stop test. Now it has
been refactored to a new function that is invoked by the test function.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Lock for controllers dict.

There are two threads running in RFMonitor, one for testing and other
for scheduling. Both the threads use controllers dictionary. So to
prevent the condition of simultaneous usage of the data structure by the
two threads, the data structure is locked when it is being used by one
of the threads and released as soon as changes are done.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer <joestringernz@gmail.com>

srijanmishra committed:
Handle testing and IPC on separate threads.

RFMonitor has two threads, on for IPC and other for testing. The earlier
architecture made IPC thread invoke testing functions. So this patch
segregates the two threads, with the main thread handling testing on the
given dictionary of controllers and IPC thread adding new entries to
dictionary when new controller spawns.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer <joestringernz@gmail.com>

srijanmishra committed:
Dict Lock, default timeout and schedule test check.

The data structure controller dictionary is being shared between IPC and
main thread, so put a lock on it when it is being used by one of them.

The timeout initial value for a test was set to 0.0 which is changed to
current time.

A flag was checked before a test was scheduled on a controller. Due to
modified control flow this condition will not arise so removed the flag
check from the schedule_test function.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer <joestringernz@gmail.com>

srijanmishra committed:
Keep count of devices connected to controllers.

To elect new master controller in case the master controller goes down,
the controller with maximum devices associated with it will be promoted.
To facilitate this, RFMonitor will maintain the count of the number of
devices associated with it.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Master election by RFMonitor.

When a master controller dies a new master is to be elected among the
remaining controllers. A dictionary of eligible masters is maintained by
the RFMonitor which is updated periodically as a new controller comes or
the load on existing controllers increase. When a master dies a master
is selected from these eligible controllers. This patch implements the
required changes at RFMonitor level to elect a new master.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
New message to elect a new master.

When a master controller dies, a new master is elected at RFMonitor and
it is to be informed to the rfproxy and inturn all the devices. So a new
message type ELECT_MASTER is created that will be used by RFMonitor to
inform the rfproxy.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer joestringernz@gmail.com

srijanmishra committed:
Inform rfproxy about new master.

When the new master is elected, it is to be informed to to rfproxy so
that it may inform the same  to all the devices. So send the ELECT_MASTER
message to rfproxy so that it may inform all the devices about the new
master.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer joestringernz@gmail.com

srijanmishra committed:
Code Cleaning.

Removed unused code from RFMonitor. Removed log message introduced for
debugging. Removed the auto code generator from msgen.py that introduced
unused import.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer joestringernz@gmail.com

srijanmishra committed:
Code Documentation.

Added doc string to all the functions and described snippets at some
places.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

srijanmishra committed:
Lock for eligible_masters

The data structure eligible_masters is being shared between IPC and main
thread. So for safe access, managed it with controller lock.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

joestringer commented:
I've sent a bunch of feedback, but there's also the issue that this code is not based against the branch that you're requesting the pull to. Please **rebase** ( _please don't merge_ ) against master if you want to push there.

srijanmishra committed:
Code Documentation.

Documents RFMonitor class attributes and fixes a doc string.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer joestringernz@gmail.com

srijanmishra committed:
Assign max count to a variable.

eligible_masters was being used directly to access max count value. For
easier usage and code readability, assigned the required value to a
variable and use it everywhere.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Joe Stringer joestringernz@gmail.com

srijanmishra commented:
@joestringer Sorry i sent the pull request to wrong branch. I have sent the request to the correct branch _ha_ and shall i close this request?

New Pull request- https://github.com/routeflow/RouteFlow/pull/4

srijanmishra committed:
Restore rftest1 and add new test for HA.

Previously the rftest1 was modified to be a test for HA. The previous
rftest1 is restored now and the modified test is moved to rftestha.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Allan Vidal <allanv@cpqd.com.br>

srijanmishra committed:
Wiki for HA and correction in rfproxy source.

A new wiki explaining new high availability module having description of
the module and architectural view of the application after incorporating
the new changes. Also corrects the rfproxys source address to the fork
in routeflow organization .

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>
Acked-by: Allan Vidal <allanv@cpqd.com.br>

srijanmishra committed:
Formatting in wiki.

The readme file had some bad formatting that is corrected.

Signed-off-by: Srijan Mishra <srijan23mishra@gmail.com>

joestringer commented:
Pull Request superceded by #4.
